github/data_source_github_release_test.go

```diff
@@ -0,0 +1,160 @@
+					resource.TestMatchResourceAttr("data.github_release.test", "url", regexp.MustCompile(`hashicorp/terraform`)),
+					resource.TestMatchResourceAttr("data.github_release.test", "tarball_url", regexp.MustCompile(`hashicorp/terraform/tarball`)),
+				),
+			},
+		},
+	})
+
+}
+
+func TestAccGithubReleaseDataSource_fetchByIdWithNoIdReturnsError(t *testing.T) {
+	retrieveBy := "id"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckGithubReleaseDataSourceConfig("", "", retrieveBy, "", 0),
+				ExpectError: regexp.MustCompile("release_id` must be set when `retrieve_by` = `id`"),
+			},
+		},
+	})
+}
+
+func TestAccGithubReleaseDataSource_fetchByIdExisting(t *testing.T) {
+	repo := "terraform"
+	owner := "hashicorp"
+	retrieveBy := "id"
+	id := int64(23055013)
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckGithubReleaseDataSourceConfig(repo, owner, retrieveBy, "", id),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.github_release.test", "release_id", strconv.FormatInt(id, 10)),
+					resource.TestMatchResourceAttr("data.github_release.test", "url", regexp.MustCompile(`hashicorp/terraform`)),
+					resource.TestMatchResourceAttr("data.github_release.test", "tarball_url", regexp.MustCompile(`hashicorp/terraform/tarball`)),
+				),
+			},
+		},
+	})
+}
+
+func TestAccGithubReleaseDataSource_fetchByTagNoTagReturnsError(t *testing.T) {
+	repo := "terraform"
+	owner := "hashicorp"
+	retrieveBy := "tag"
+	id := int64(23055013)
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckGithubReleaseDataSourceConfig(repo, owner, retrieveBy, "", id),
+				ExpectError: regexp.MustCompile("`release_tag` must be set when `retrieve_by` = `tag`"),
+			},
+		},
+	})
+}
+
+func TestAccGithubReleaseDataSource_fetchByTagExisting(t *testing.T) {
+	repo := "terraform"
+	owner := "hashicorp"
+	retrieveBy := "tag"
+	tag := "v0.12.20"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckGithubReleaseDataSourceConfig(repo, owner, retrieveBy, tag, 0),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.github_release.test", "release_tag", tag),
+					resource.TestMatchResourceAttr("data.github_release.test", "url", regexp.MustCompile(`hashicorp/terraform`)),
+					resource.TestMatchResourceAttr("data.github_release.test", "tarball_url", regexp.MustCompile(`hashicorp/terraform/tarball`)),
+				),
+			},
+		},
+	})
+}
```

Comment:
The convention within this repository I've observed so far has been to not hard-code values such as the repository and tag as was done in this test.  I think this test is stable but would like to see an attempt to obtain these values from the environment before proceeding with the hard-coded defaults.  See [here](https://github.com/terraform-providers/terraform-provider-github/pull/309#discussion_r362785464) for prior art.

Reply:
Thanks for referring to this. I was unsure how to proceed since we don't have an associated resource to create / delete a release. Have a look at my changes - adding another environment variable with the tag and sourcing it from the template repo.

In future, if added, I think it would be good to manage this release test object via a release resource as well, making it fully isolated.